File: 1021.cpp
#include <cstdio>
#include <cstring>
#include <queue>
#include <algorithm>
using namespace std;
int ls[20005],tr[20005],nx[20005],dep[20005],newdep[20005],tot = 0;
bool inAns[20005];
int color[20005];
int atail = 0,Ans[20005];
void Add(int a,int b)
{
	nx[++tot] = ls[a];
	ls[a] = tot;
	tr[tot] = b;
}
queue<int> q;
int main()
{
	int i,j,n,a,b,t,cs = 0,longest = 0,longest2 = 0;
	scanf("%d",&n);
	for (i = 1;i < n;++i)
	{
		scanf("%d%d",&a,&b);
		Add(a,b); Add(b,a);
	}
	for (i = 1;i <= n;++i)
		if (!color[i])
		{
			color[i] = ++cs;
			q.push(i);
			while (!q.empty())
			{
				t = q.front(); q.pop();
				for (j = ls[t];j;j = nx[j])
					if (!color[tr[j]])
					{
						color[tr[j]] = cs;
						dep[tr[j]] = dep[t] + 1;
						longest = longest > dep[tr[j]] ? longest : dep[tr[j]];
						q.push(tr[j]);
					}
			}
		}
	if (cs > 1)
	{
		printf("Error: %d components\n", cs);
		return 0;
	}
	for (i = 1;i <= n;++i)
		if (dep[i] == longest)
		{
			if (!inAns[i])
			{
				Ans[++atail] = i;
				inAns[i] = true;
			}
			memset(newdep, -1, sizeof(newdep));
			q.push(i);
			newdep[i] = 0;
			while (!q.empty())
			{
				t = q.front(); q.pop();
				for (j = ls[t];j;j = nx[j])
					if (newdep[tr[j]] == -1)
					{
						newdep[tr[j]] = newdep[t] + 1;
						longest2 = longest2 > newdep[tr[j]] ? longest2 : newdep[tr[j]];
						q.push(tr[j]);
					}
			}
			for (j = 1;j <= n;++j)
				if (newdep[j] == longest2 && !inAns[j])
				{
					Ans[++atail] = j;
					inAns[j] = true;
				}
		}
	sort(Ans + 1,Ans + atail + 1);
	for (i = 1;i <= atail;++i)
		printf("%d\n",Ans[i]);
	return 0;
}

File: 1021.rb
# TLE
n = gets.chomp.to_i
edge = Array.new(n) { Array.new }
(n - 1).times do
    a, b = gets.chomp.split(" ").collect { |x| x.to_i - 1 }
    edge[a] << b
    edge[b] << a
end
color = []; dep = Array.new(n)
q = []
csum = 0; long = 0
0.upto(n - 1) { |x|
    if !color[x]
        csum += 1
        color[x] = csum
        dep[x] = 0
        q << x
        while !q.empty? do
            t = q.shift
            edge[t].each { |tr|
                if !color[tr]
                    color[tr] = csum
                    dep[tr] = dep[t] + 1
                    long = long > dep[tr] ? long : dep[tr]
                    q << tr
                end
            }
        end
    end
}
if csum != 1
    puts "Error: #{csum} components"
    exit
end
ans = []
dep.each_with_index { |x,i|
    if x == long
        longest = 0
        ans << i
        q.clear; q << i
        newdep = Array.new(n); newdep[i] = 0
        while !q.empty? do
            t = q.shift
            edge[t].each { |tr|
                if !newdep[tr]
                    newdep[tr] = newdep[t] + 1
                    longest = longest > newdep[tr] ? longest : newdep[tr]
                    q << tr
                end
            }
        end
        newdep.each_index { |new| ans << new if newdep[new] == longest}
    end
}
puts ans.uniq.sort.collect { |x| x + 1 }.join("\n")

